patches/api/0387-Paper-Plugins.patch

```diff
@@ -0,0 +1,2413 @@
+new file mode 100644
+index 0000000000000000000000000000000000000000..82d387b55a25e5418c763b97b77f5bec93f86e41
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/plugin/provider/service/classloader/PaperPluginClassLoaderStorage.java
+@@ -0,0 +1,81 @@
++package io.papermc.paper.plugin.provider.service.classloader;
++
++import io.papermc.paper.plugin.configuration.PluginConfiguration;
++import org.bukkit.Bukkit;
++import org.bukkit.plugin.PluginManager;
++import org.jetbrains.annotations.ApiStatus;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.HashMap;
++import java.util.List;
++import java.util.Map;
++import java.util.concurrent.CopyOnWriteArrayList;
++import java.util.concurrent.locks.ReentrantReadWriteLock;
++
++/**
++ * This is used for connecting multiple classloaders.
++ */
++@ApiStatus.Internal
++public final class PaperPluginClassLoaderStorage {
++
++    private static final boolean DISABLE_CLASS_PRIORITIZATION = Boolean.getBoolean("Paper.DisableClassPrioritization");
++
++    private final Map<String, ReentrantReadWriteLock> classLoadLock = new HashMap<>();
++    private final Map<String, Integer> classLoadLockCount = new HashMap<>();
++    public final List<ConfiguredPluginClassLoader> classLoaders = new CopyOnWriteArrayList<>();
++
++    public static final PaperPluginClassLoaderStorage INSTANCE = new PaperPluginClassLoaderStorage();
++
++    public PaperPluginClassLoaderStorage() {
++    }
++
++    public Class<?> getClassByName(String name, boolean resolve, PluginConfiguration pluginConfiguration, @Nullable ConfiguredPluginClassLoader requester) {
++        // make MT safe
++        java.util.concurrent.locks.ReentrantReadWriteLock lock;
++        synchronized (this.classLoadLock) {
++            lock = this.classLoadLock.computeIfAbsent(name, (x) -> new java.util.concurrent.locks.ReentrantReadWriteLock());
++            this.classLoadLockCount.compute(name, (x, prev) -> prev != null ? prev + 1 : 1);
++        }
++        lock.writeLock().lock();
++
++        try {
++            PluginManager manager = Bukkit.getPluginManager();
++            if (!DISABLE_CLASS_PRIORITIZATION && requester != null) {
++                try {
++                    return requester.loadClass(name, false, false, manager.isTransitiveDependency(pluginConfiguration, requester.getConfiguration()));
++                } catch (ClassNotFoundException cnfe) {
++                }
++            }
++            for (ConfiguredPluginClassLoader loader : this.classLoaders) {
++                try {
++                    return loader.loadClass(name, resolve, false, manager.isTransitiveDependency(pluginConfiguration, loader.getConfiguration()));
++                } catch (ClassNotFoundException cnfe) {
++                }
++            }
```

Comment:
There might be an NPE here. Imagine a situation where we have two PluginBootstraps. PluginBootstrap B depends on PluginBootstrap A. There is a "MiscUtils" class in the same Jar with PluginBootstrap A, and PluginBootstrap A uses MiscUtils in its `.bootstrap()` method. So both PluginBootstrap A and MiscUtils are loaded in the same PaperPluginClassLoader. But if PluginBootstrap B also uses MiscUtils from the PluginBootstrap A jar, then ClassNotFoundException occurs in `PaperPluginClassLoader.loadClass(String name, boolean resolve, boolean checkGlobal, boolean checkLibraries)` for some reason:
```java
        try {
            Class<?> result = super.loadClass(name, resolve);

            // SPIGOT-6749: Library classes will appear in the above, but we don't want to return them to other plugins
            if (checkGlobal || result.getClassLoader() == this) {
                return result;
            }
        } catch (ClassNotFoundException ex) {
        }
```
So after this code it checks for all dependencies and in while checking "global" dependecnies it uses `PaperPluginClassLoaderStorage.getClassByName(String name, boolean resolve, PluginConfiguration pluginConfiguration, @Nullable ConfiguredPluginClassLoader requester)` and inside this method it checks for transitive dependencies using `PluginManager.isTransitiveDependency(PluginConfiguration pluginConfiguration, PluginConfiguration dependencyConfig)` but at this stage Bukkit API hasn't loaded yet, so it cannot get PluginManager instance, so it all throws something like this:
```java
Sep 03, 2022 4:54:36 PM io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.BootstrapProviderStorage$1 load
SEVERE: Failed to run bootstrapper for /home/denery/IdeaProjects/Toky/toky-personality/build/libs/toky-personality-unspecified.jar. This plugin will not be loaded.
java.lang.NullPointerException: Cannot invoke "org.bukkit.Server.getPluginManager()" because "org.bukkit.Bukkit.server" is null
	at org.bukkit.Bukkit.getPluginManager(Bukkit.java:723)
	at io.papermc.paper.plugin.provider.service.classloader.PaperPluginClassLoaderStorage.getClassByName(PaperPluginClassLoaderStorage.java:42)
	at io.papermc.paper.plugin.provider.service.classloader.PaperPluginClassLoader.loadClass(PaperPluginClassLoader.java:87)
	at io.papermc.paper.plugin.provider.service.classloader.PaperPluginClassLoader.loadClass(PaperPluginClassLoader.java:57)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at toky-personality-unspecified.jar//io.toky.personality.PersonalityKernel.boostrap(PersonalityKernel.java:33)
	at io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.BootstrapProviderStorage$1.load(BootstrapProviderStorage.java:46)
	at io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.BootstrapProviderStorage$1.load(BootstrapProviderStorage.java:34)
	at io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.strategy.ModernPluginLoadingStrategy.loadProviders(ModernPluginLoadingStrategy.java:126)
	at io.papermc.paper.plugin.provider.service.entrypoint.storage.ordered.OrderedProviderStorage.enter(OrderedProviderStorage.java:25)
	at io.papermc.paper.plugin.provider.service.entrypoint.ServerEntryPointHandler.enter(ServerEntryPointHandler.java:33)
	at net.minecraft.server.Main.main(Main.java:127)
	at org.bukkit.craftbukkit.Main.main(Main.java:300)
	at io.papermc.paperclip.Paperclip.lambda$main$0(Paperclip.java:42)
	at java.base/java.lang.Thread.run(Thread.java:833)
```
The solution is to move the `.isTransitiveDependency()` method somewhere else which doesn't use Bukkit API and has dependencyGraph. (to the ProviderLoaderStrategy which also builds its own dependencyGraph while loading Providers?)

Reply:
Oh good catch, yeah I need to properly pass a dependency context there. Thanks!

Reply:
Thank you very much for this! I have resolved this as of https://github.com/PaperMC/Paper/pull/8108/commits/789da71b5769990915ade219f6f05da8947708c7

Feel free to take a look :)